### ISSUE TITLE
`fluent-plugin-loggly` requires an older version of `net-http-persistent`

### ISSUE THREAD

**Issue by gg-kialo** [vmware/kube-fluentd-operator#221]
In `v1.15.0` the bundled `fluent-plugin-loggly@v0.1.0-pre` does not work with the bundled `net-http-persistent@v3.1.0`

`fluent-plugin-loggly` expects a pre-3.0.0 version of `net-http-persistent`. In the `3.0.0` releases, `Net::HTTP::Persistent.new` has been changed to accept keyword arguments instead of optional positional arguments.

This causes the validation of fluent configs to fail. E.g. when running `validate-from-dir.sh`:

```
*** Error for namespace 'kube-system': 2021-06-24 09:46:12 +0000 [error]: unexpected error error_class=ArgumentError error="wrong number of arguments (given 2, expected 0)"
  2021-06-24 09:46:12 +0000 [error]: /usr/local/lib/ruby/gems/2.7.0/gems/net-http-persistent-3.1.0/lib/net/http/persistent.rb:530:in `initialize'
  2021-06-24 09:46:12 +0000 [error]: /usr/local/lib/ruby/gems/2.7.0/gems/fluent-plugin-loggly-0.1.0.pre/lib/fluent/plugin/out_loggly_buffered.rb:43:in `new'
```

**Comment by gg-kialo** [comment]
I'd be interested in knowing any workarounds, otherwise, I will continue to use `v1.13.0`

**Comment by Cryptophobia** [comment]
Thank you for reporting @gg-kialo. This likely broke because `net-http-persistent` gem was updated automatically to 3.1.0 when we built `v1.15.0`, but the `fluent-plugin-loggly` does not specify that it does not accept `< 3.0.0` of `net-http-persistent` gem.

```
gem dependency fluent-plugin-loggly --reverse-dependencies
Gem fluent-plugin-loggly-0.1.0.pre
  net-http-persistent (>= 2.7)
  yajl-ruby (>= 1.0
```

After looking more into this, looks to be a versioning issue. Tag `v0.0.9` actually has the fix on top of `v0.1.0.pre`. 
https://github.com/patant/fluent-plugin-loggly/compare/v0.1.0.pre...v0.0.9

This gem looks that it has not been supported since 2016 (5 years). We will use tag v0.0.9 for v1.15.1 release of KFO, but it might be advisable to find a replacement gem or deprecate this gem.

**Comment by Cryptophobia** [comment]
Unfortunately, two other gems depend on upgraded versions of `net-http-persistent` gem that fixes connectivity issues for those gems:

```
Bundler could not find compatible versions for gem "net-http-persistent":
  In Gemfile:
    fluent-plugin-datadog (= 0.13.0) was resolved to 0.13.0, which depends on
      net-http-persistent (~> 3.1)

    fluent-plugin-logzio (= 0.0.20) was resolved to 0.0.20, which depends on
      net-http-persistent (~> 2.9)
```

It would be unfair if we were to downgrade those gems in order to run a gem that has not been updated in 5+ years. Can you open an issue with the maintainers?

**Comment by gg-kialo** [comment]
> Can you open an issue with the maintainers?

I will do, but I don't expect anything from it!

**Comment by Cryptophobia** [comment]
> > Can you open an issue with the maintainers?
> 
> I will do, but I don't expect anything from it!

You can also try to fork the gem and fix it yourself. We will pull version from your fork. It believe one line change like this one may solve your issues to use the new client for net-http gem:

https://github.com/DataDog/fluent-plugin-datadog/commit/f3a4b428f4c61158094fb38ea16dd7b6107f309c#diff-8abf650fc7e6ac50d9d48da76f3c7cadda95f72a285897c45021c287b4d364a5L305-R309

**Comment by Cryptophobia** [comment]
Two lines need to change: https://github.com/patant/fluent-plugin-loggly/search?q=Net%3A%3AHTTP%3A%3APersistent.new
And then two more lines in .gemspec files, you increment version to `0.1.1`, and change to accept `s.add_dependency('net-http-persistent', '~> 3.0')` [here](https://github.com/patant/fluent-plugin-loggly/blob/master/fluent-plugin-loggly.gemspec#L21) .

**Comment by Cryptophobia** [comment]
@gg-kialo , I applied the fixes above on a forked repo (https://github.com/Cryptophobia/fluent-plugin-loggly) and it starts up correctly with minimal efforts. This may be fixed and working in `v1.15.1`. Please try again when v1.15.1 is released. 

```
root [ /fluentd ]# /usr/local/bin/fluentd --no-supervisor --dry-run -p /fluentd/plugins/ -c test.conf
2021-06-28 16:46:31 +0000 [info]: parsing config file is succeeded path="test.conf"
2021-06-28 16:46:31 +0000 [info]: starting fluentd-1.13.1 without supervision pid=21 ruby="2.7.3"
2021-06-28 16:46:31 +0000 [warn]: 'type' is deprecated parameter name. use '@type' instead.
2021-06-28 16:46:31 +0000 [info]: adding match pattern="your_match" type="loggly_buffered"
2021-06-28 16:46:31 +0000 [info]: using configuration file: <ROOT>
  <match your_match>
    type loggly_buffered
    loggly_url "https://logs-01.loggly.com/bulk/xxx-xxxx-xxxx-xxxxx-xxxxxxxxxx"
    output_include_time true
    time_precision_digits 3
    buffer_type "file"
    buffer_path /path/to/buffer/file
    flush_interval 10s
    <buffer>
      flush_mode interval
      retry_type exponential_backoff
      @type file
      path /path/to/buffer/file
      flush_interval 10s
    </buffer>
  </match>
</ROOT>
2021-06-28 16:46:31 +0000 [info]: starting fluentd worker pid=21 ppid=1 worker=0
2021-06-28 16:46:31 +0000 [info]: fluentd worker is now running worker=0
```


However, if this gem does not get updated, we may have to deprecate support and I made a note for the future.
